### PR TITLE
Extend support for git cache to all strategies

### PR DESCRIPTION
**Pull request by mortent**
This makes several improvements to how kpt uses git during `get` and `update`:
 * Leverage the cache dir to avoid having to fetch all refs and objects for a repo every time.
 * Fetch refs and the commit their pointing at as the first request to the upstream git repo:
   * Allows us to check for existence of package tags without trying to fetch the ref (and its objects) and therefore avoid what is now duplicated effort if the package ref doesn't exist and we have to fall back to the repo-level ref.
   * Since we will always have the commit SHA for every ref, we can always do `git reset --hard` to check out the specific commit. This means we can avoid creating any local refs (i.e. do `git checkout`).
   * The `git ls-remote` command we use to fetch refs has few failure scenarios if we can talk to the repo. So an error from this command almost certainly means we can't talk to the repo (so auth issues, doesn't exist...).

Some consequences of this change:
 * We no longer fetch git submodules. I think this a case that is unlikely to be used much (and if I'm wrong, we can add support for this later) and we already provide subpackages to handle this kind of cases.
 * This doesn't yet have a way to automatically clean up the kpt cache. Probably not a huge concern for now, but we should see if there is a good way automatically prune.

Fixes: https://github.com/GoogleContainerTools/kpt/issues/1618, https://github.com/GoogleContainerTools/kpt/issues/1376, https://github.com/GoogleContainerTools/kpt/issues/1527, https://github.com/GoogleContainerTools/kpt/issues/1079

### CODE REVIEW

File: internal/util/fetch/fetch.go [hunk 1]
```diff
@@ -127,142 +129,46 @@ func cloneAndCopy(ctx context.Context, r *git.RepoSpec, dest string) error {
...
-	// look for a tag with the directory as a prefix for versioning
-	// subdirectories independently
-	originalRef := repoSpec.Ref
-	if repoSpec.Path != "" && !strings.Contains(repoSpec.Ref, "refs") {
-		// join the directory with the Ref (stripping the preceding '/' if it exists)
-		repoSpec.Ref = path.Join(strings.TrimLeft(repoSpec.Path, "/"), repoSpec.Ref)
-	}
 
-	defaultRef, err := gitutil.DefaultRef(repoSpec.OrgRepo)
+	upGitRunner, err := gitutil.NewGitUpstreamRepo(ctx, repoSpec.CloneSpec())
```
Comment (frankfarzan) [inline internal/util/fetch/fetch.go:133]: naming: not sure what `up` refers to in the name at first pass, consider renaming or using a short name if small scope. Suggestions:

upstreamGit
upstream
up

File: internal/util/fetch/fetch.go [hunk 2]
```diff
@@ -127,142 +129,46 @@ func cloneAndCopy(ctx context.Context, r *git.RepoSpec, dest string) error {
...
-		}
-
-		err = runGitExec(ctx, repoSpec.Dir, "reset", "--hard", "FETCH_HEAD")
-		if err != nil {
-			return errors.E(op, errors.Git,
-				fmt.Errorf("trouble hard resetting empty repository to %s: %w", repoSpec.Ref, err))
-		}
-		return nil
-	}()
+	_, err = gitRunner.Run(ctx, "reset", "--hard", commit)
```
Comment (frankfarzan) [inline internal/util/fetch/fetch.go:158]: can you add comments to any non-obvious section of the code. For example, why are you running `reset --hard`.

File: internal/util/fetch/fetch.go (hunk 1 above)
Comment (mortent) [reply internal/util/fetch/fetch.go]: Renamed this to `upstreamRepo`. The variable name was a leftover from before a refactor.

File: internal/util/fetch/fetch.go (hunk 2 above)
Comment (mortent) [reply internal/util/fetch/fetch.go:170]: I have added more comments about what we are doing here.

File: internal/gitutil/gitutil.go [hunk 3]
```diff
@@ -32,144 +36,282 @@ import (
...
+			return commit, true
+		}
+	}
+	return "", false
+}
+
+// ResolveRef resolves the ref (either tag or branch) to a commit SHA. If the
+// ref doesn't exist in the upstream repo, the last return value will be false.
+func (gur *GitUpstreamRepo) ResolveRef(ref string) (string, bool) {
+	commit, found := gur.ResolveBranch(ref)
```
Comment (phanimarupaka) [inline internal/gitutil/gitutil.go:299]: Any reason for this precedence ? If ref and branch has same name we pull the branch ?

**Comment by phanimarupaka** [comment]
> This makes several improvements to how kpt uses git during `get` and `update`:
> 
> * Leverage the cache dir to avoid having to fetch all refs and objects for a repo every time.
> * Fetch refs and the commit their pointing at as the first request to the upstream git repo:
>   
>   * Allows us to check for existence of package tags without trying to fetch the ref (and its objects) and therefore avoid what is now duplicated effort if the package ref doesn't exist and we have to fall back to the repo-level ref.
>   * Since we will always have the commit SHA for every ref, we can always do `git reset --hard` to check out the specific commit. This means we can avoid creating any local refs (i.e. do `git checkout`).
>   * The `git ls-remote` command we use to fetch refs has few failure scenarios if we can talk to the repo. So an error from this command almost certainly means we can't talk to the repo (so auth issues, doesn't exist...).
> 
> Some consequences of this change:
> 
> * We no longer fetch git submodules. I think this a case that is unlikely to be used much (and if I'm wrong, we can add support for this later) and we already provide subpackages to handle this kind of cases.
> * This doesn't yet have a way to automatically clean up the kpt cache. Probably not a huge concern for now, but we should see if there is a good way automatically prune.
> 
How are cache repos cleaned currently ? For every get/update command, the previous repo is cleaned ?
> Fixes: #1618, #1376, #1527, #1079

**Comment by mortent** [comment]
> > * This doesn't yet have a way to automatically clean up the kpt cache. Probably not a huge concern for now, but we should see if there is a good way automatically prune.
> 
> How are cache repos cleaned currently ? For every get/update command, the previous repo is cleaned ?

Current version of kpt also doesn't prune the cache. 
We don't remove the cache between every get/update command. This means that when we do `git fetch` command on subsequent get/update, even if there are updates, we only need to fetch new objects instead of the full tree.

**Review by phanimarupaka — APPROVED** [review]

**Pull request merged**